rflx/pyrflx/typevalue.py

```diff
@@ -492,42 +504,69 @@ class MessageValue(TypeValue):
 
     _type: Message
 
-    def __init__(self, model: Message, refinements: Sequence[Refinement] = None) -> None:
+    def __init__(
+        self,
+        model: Message,
+        refinements: Sequence[Refinement] = None,
+        skip_verification: bool = False,
+        fields: Mapping[str, "MessageValue.Field"] = None,
+        checksums: Mapping[str, "MessageValue.Checksum"] = None,
+        type_literals: Mapping[Name, Expr] = None,
+    ) -> None:
+        # pylint: disable=too-many-arguments
         super().__init__(model)
+        self._skip_verification = skip_verification
         self._refinements = refinements or []
-        self._fields: Dict[str, MessageValue.Field] = {
+
+        self._fields: Mapping[str, MessageValue.Field] = fields if fields else {
             f.name: self.Field(
                 TypeValue.construct(
-                    self._type.types[f], imported=self._type.types[f].package != model.package
+                    self._type.types[f] if f in self._type.types else Opaque(),
+                    imported=f in self._type.types and self._type.types[f].package != model.package,
                 ),
                 f.name,
             )
-            for f in self._type.fields
+            for f in (INITIAL,) + self._type.fields
         }
 
-        self._checksums: Mapping[str, MessageValue.Checksum] = {
+        self._checksums: Mapping[str, MessageValue.Checksum] = checksums if checksums else {
             str(field_name): MessageValue.Checksum(str(field_name), parameters)
             for field_name, parameters in self._type.checksums.items()
         }
 
-        self.__type_literals: Mapping[Name, Expr] = {}
-        self._last_field: str = self._next_field(INITIAL.name)
-        for t in [
-            f.typeval.literals for f in self._fields.values() if isinstance(f.typeval, EnumValue)
-        ]:
-            self.__type_literals.update(t)
-        self.__type_literals_keys = set(self.__type_literals)
-        initial = self.Field(OpaqueValue(Opaque()), INITIAL.name)
+        self.__type_literals: Mapping[Name, Expr] = type_literals if type_literals else {}
+        if not self.__type_literals:
+            assert isinstance(self.__type_literals, dict)
```

Comment:
This assertion shouldn't be necessary. I suppose you only need it because you are using an immutable mapping as type.

Reply:
The `Mapping` is used since the type literals should not change. If they are not set and have to be filled this has to be done as a `dict` since `update` cannot be called on mappings. If this assertion is not set https://github.com/Componolit/RecordFlux/blob/fde3766d2cea062d711d8c584c8b1a16a1f03093/rflx/pyrflx/typevalue.py#L545 will fail a type check.

Reply:
I'm aware of that, but this solution is a hack. If you want to have an immutable mapping, you should use a list comprehension for initialization. I see no advantage in using a for-loop and `update` here.

Reply:
Fixed in https://github.com/Componolit/RecordFlux/pull/405/commits/17a6efab4e5ad6edb8dfd1c5020ffee5cd54dca5.